qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py

```diff
@@ -336,13 +342,16 @@ def mcrz(
     self._check_dups(all_qubits)
 
     n_c = len(control_qubits)
-    if n_c == 1:  # cu
-        _apply_cu(self, 0, 0, lam, control_qubits[0], target_qubit, use_basis_gates=use_basis_gates)
+    if n_c == 1:
+        self.append(CRZGate(lam), control_qubits + [target_qubit])
     else:
-        lam_step = lam * (1 / (2 ** (n_c - 1)))
-        _apply_mcu_graycode(
-            self, 0, 0, lam_step, control_qubits, target_qubit, use_basis_gates=use_basis_gates
-        )
+        mcsu2_real_diagonal(self, RZGate(lam).to_matrix(), control_qubits, target_qubit)
+
+    if use_basis_gates:
+        # pylint: disable=cyclic-import
+        from qiskit import transpile
+
+        self = transpile(self, basis_gates=["p", "u", "cx"], optimization_level=0)
```

Comment:
I don't think this is quite correct, sorry if I didn't express it correctly before. With this code, the whole circuit will get transpiled to the specified basis -- but we only want the newly added mcrz to be in that basis 🙂 maybe we can just change `mcsu2_real_diagonal` to return the MCRZ subcircuit and then transpile that?

Reply:
Hi @Cryoris , I removed the transpile call. 
I think that we should not transpile the MCRZ circuit here, because of the transpile cost with a higher number of controls.

Reply:
I agree it's not the most efficient way, but we should ensure that `use_basis_gates=True` will still behave as it did before 🙂 If we want to avoid transpiling, would it be possible to express `mcsu2` in the `P U CX` basis?